Perf: Implement branchless compare

thinkbeforecoding opened the pull request:
This is an implementation of #13098 
It implements a branchless compare:
cgt x y - clt x y
When x > y: 1 - 0 = 1
When x < y: 0 - 1 = -1
When x = y: 0 - 0 = 0

Benchmarks show that it is very slightly slower (10%) when predictions are always correct, but it is 3x faster on random values where prediction fail more often

The code emitted by the JIT is the same size.

thinkbeforecoding commented:
Some IL tests are failing due to the change.

There are some build errors on my machine (it seems to be new F# features like [x..] indexing... How can I build this locally ?

vzarytovskii commented:
> Some IL tests are failing due to the change.
> 
> There are some build errors on my machine (it seems to be new F# features like [x..] indexing... How can I build this locally ?
> 

Build scripts (from the devguide) should work just fine. You can use it with noVisualStudio switch.

Devguide also describes how to update IL baselines (if test uses baseline files).

dsyme commented:
I'm curious if we know that this is the fastest implementation.  Is there any information on this on the web e.g. for C++ or assembly code?

dsyme commented:
I'd like to see a systematic test matrix for comparisons on all the basic types affected by this, including 

* MaxValue
* MinValue
* 0
* 1
* -1

dsyme commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -4771,17 +4771,17 @@ namespace Microsoft.FSharp.Core
-                 when ^T : uint32 = if (# "clt.un" e1 e2 : bool #) then (-1) else (# "cgt.un" e1 e2 : int #)
-                 when ^T : uint64 = if (# "clt.un" e1 e2 : bool #) then (-1) else (# "cgt.un" e1 e2 : int #)
-                 when ^T : unativeint = if (# "clt.un" e1 e2 : bool #) then (-1) else (# "cgt.un" e1 e2 : int #)
+                 when ^T : bool   = (# "cgt" e1 e2 : int #) - (# "clt" e1 e2 : int #)
```
I think the `cgt` and `clt` are unnecessary as `bool` is spec'd as 0 or 1

thinkbeforecoding replied:
True. 😁

dsyme replied:
I guess you still have to get 0, 1, - 1 somehow :) There must be some magic to do that. Maybe cgt/clt is good enough

thinkbeforecoding replied:
In IL True is 1 and False is 0 integer
Son
True > False 1 - 0 = 1
False < True 0 - 1 = - 1
True = True 1-1 = 0
False = False 0-0 = 0

So a simple subtraction x-y is enough

dsyme commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -1137,17 +1137,17 @@ namespace Microsoft.FSharp.Core
-                 when 'T : uint32 = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
-                 when 'T : uint64 = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
-                 when 'T : unativeint = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
+                 when 'T : bool   = (# "cgt" x y : int #) - (# "clt" x y : int #)
```
I think the `cgt` and `clt` are unnecessary as `bool` is spec'd as 0 or 1

thinkbeforecoding commented:
There could be some improvements in the jit.
For instance, the cmp is done twice.
Since setgt setlt don't modify flags, both could be done with a single cmp.
And the double zero extension could be done after the sub as a sign extension.

thinkbeforecoding commented:
I think the fastest version would be in x64
cmp eax ecx
setg al
setl dl
sub dl, al
movsx eax, dl

There is a 4 instr version found on SO

sub %1, %0
jno 1f
 cmc
 rcr %0
https://stackoverflow.com/questions/10996418/efficient-integer-compare-function

But it involve a conditional jump that will be slower on branch prediction. Moreover I see no way to make the jit compile something like this.

thinkbeforecoding commented:
I was also looking at movcc (conditional move) instructions, but the source is a register or memory, it cannot be a constant. So it requires extra instructions to load constants and won't we shorter. 
That could be interesting to implement min and max for instance.

thinkbeforecoding commented:
After a check, the requirement on IComparable<T>.CompareTo is:
* when x > y, result is > 0
* when x < y, result is < 0
* when x=y result is 0

for shorter types (byte, short, char) it is implemented as: 
```
int x - int y
```
This is valid since there  is no risk of overflow. But the result is not necessarily 0,1,-1

For longer types, it uses conditional jumps and returns always 0,1,-1

Do we have that constraint on the compare function ?

vzarytovskii commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -1059,14 +1059,22 @@ namespace Microsoft.FSharp.Core
-                 when 'T : int16  = (# "cgt" x y : int #) - (# "clt" x y : int #)
+
+                 when 'T : bool   =
+                    // bool is alread 1 for True or 0 for False
```
```suggestion
                    // bool is already 1 for True or 0 for False
```

vzarytovskii commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -1137,14 +1145,22 @@ namespace Microsoft.FSharp.Core
-                 when 'T : int16  = (# "cgt" x y : int #) - (# "clt" x y : int #)
+                    
+                 when 'T : bool   = 
+                    // bool is alread 1 for True or 0 for False
```
```suggestion
                    // bool is already 1 for True or 0 for False
```

thinkbeforecoding replied:
😁

vzarytovskii commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -4771,14 +4787,21 @@ namespace Microsoft.FSharp.Core
-                 when ^T : sbyte  = (# "cgt" e1 e2 : int #) - (# "clt" e1 e2 : int #)
-                 when ^T : int16  = (# "cgt" e1 e2 : int #) - (# "clt" e1 e2 : int #)
+                 when ^T : bool   =
+                    // bool is alread 1 for true or 0 for false
```
```suggestion
                    // bool is already 1 for true or 0 for false
```

vzarytovskii commented on tests/fsharp/Compiler/Language/ComparisonOptimizationTest.fs:
```diff
@@ -0,0 +1,381 @@
+﻿namespace FSharp.Compiler.UnitTests
```
It's fine in this PR, but for future, we suggest creating new tests in `FSharp.Compiler.ComponentTests`, it has nicer APIs and better support for IL baseline tests (instead of having an inline IL, it can be a separate file, which does not require recompiling the suite in case of change, and it is easier to re-generate using the env variable).

thinkbeforecoding commented:
There are a lot of culture dependent tests that fail on my french machine . Fixed it.
There was also a lot of over specific autogenerated tests for compare. They were checking for the exact response of compare... The check must be done on the sign, not the exact value. Fixed it.

thinkbeforecoding commented:
Some of the builds fail but it doesn't seem related to the change.. Any idea?

vzarytovskii commented:
> Some of the builds fail but it doesn't seem related to the change.. Any idea?

The latest run failed because of the tests:
https://dev.azure.com/dnceng/public/_build/results?buildId=1791081&view=ms.vss-test-web.build-test-results-tab&runId=47885210&resultId=100293&paneView=debug

thinkbeforecoding commented:
These are baseline tests, but they don't seem to be updated with the instructions above...

vzarytovskii commented:
I've updated them + updated the guide, If I may ask, which shell were you using?

vzarytovskii commented:
Great, now it's even more failures...I will take a look.

thinkbeforecoding commented:
I was using powershell so I used
$env:TEST_UPDATE_BSL=1

vzarytovskii commented:
> I was using powershell so I used
> $env:TEST_UPDATE_BSL=1

Hm, interesting. I will probably add a separate switch to the build scripts which runs tests and updates baselines.

vzarytovskii commented:
@thinkbeforecoding I have updated 2 remaining baselines for net472 framework + updated devguide describing how it works.
To update them successfully, It requires some knowledge of how IL baselines work. I will add a separate command to build scripts for updating those (https://github.com/dotnet/fsharp/issues/13204)

The rest of the tests which are failing are checking the compare result.

KevinRansom commented:
> @thinkbeforecoding I have updated 2 remaining baselines for net472 framework + updated devguide describing how it works. To update them successfully, It requires some knowledge of how IL baselines work. I will add a separate command to build scripts for updating those (#13204)
> 
> The rest of the tests which are failing are checking the compare result.

computers are harsh:
System.Exception : args({ Item = 0us }, { Item = 65535us }) Expected=-1. Received=-65535.
0 - 65535 is indeed -65535 and not the -1 the program was expecting.

KevinRansom commented:
> Since compare fallbacks to ICompare.Compare for other types, there is no guarantee that the output is always 0,1,-1.

I don't know about the no guarantee thing.  If someone upgrades their dotnet sdk recompiles their app, and their original source code now has a "hard to analyze bug" pretty sure we would revert the change whilst looking red faced when saying yes we knew it could break people, but we thought it was rare enough to risk.  And amusingly it was even caught by tests, which is an eye opener.

thinkbeforecoding commented:
Actually, byte.CompareTo or char.CompareTo return a different result from F# compare. Some types return only 0,-1,1 others return a large range of values.

Yet I can understand the impact at large scale.
However, the Bcl decided to randomize GetHashCode which was a similar breaking change. The spec has always specified not to rely on it value, but it probably happened... Was there bad feedback?

Would it be possible to add an analyser or a warning for comparison of the result of compare to anything other than 0?

thinkbeforecoding commented:
For now, progress is hindered by https://github.com/dotnet/fsharp/issues/13212  Running test suites locally is problematic on non en-US machines.

thinkbeforecoding commented on src/FSharp.Core/prim-types.fs:
```diff
@@ -1137,17 +1145,25 @@ namespace Microsoft.FSharp.Core
-                 when 'T : unativeint = if (# "clt.un" x y : bool #) then (-1) else (# "cgt.un" x y : int #)
+                    
+                 when 'T : bool   = 
+                    // bool is already 1 for True or 0 for False
```
I have to check the specification of the IL but as far as I know, 1 and 0 are the only legal value..

KevinRansom replied:
I did some tests, and indeed the CLR does marshal, non-zero to bool as 1.

KevinRansom commented:
Okay this is a great change, however, I am very concerned about compatibility.  The optimizations for bool, byte, sbyte, char, int16 and uint16 all return values < 0 instead of -1 and > 0 instead of 1.  And so are detectable.

The FSharp.Core unit test library has a number of api tests verifying the behaviour of compare.  108 of them fail because of the change in the return value of compare.

I took a look at your changes and we can get most of the performance benefits without the compatibility break. and be branchless on compare.

For int32, int64, uint32 etc. this PR generates :
 (# "cgt" x y : int #) - (# "clt" x y : int #)
Which very nicely returns -1, 0 or 1.

I added a Benchmark for each of the types: bool, byte, sbyte, char, int16 and uint16 with the same generated code and ran the benchmarks.

StdType         is what exists in the rtm product
FastType        is the code that is in this pr.
OtherType     is the compatible branchless version

Bool comparison code ---
````
let inline stdcmpBool (x: bool) (y: bool) =
    if (# "clt" x y : bool #) then (-1) else (# "cgt" x y : int #)

let inline fastcmpBool (x: bool) (y: bool) =
    (# "" x : int #) - (# "" y : int #)

let inline othercmpBool (x: bool) (y: bool) =
    (# "cgt" x y : int #) - (# "clt" x y : int #) 
````

Results:
````
|    Method |     Mean |    Error |   StdDev | Ratio | RatioSD |
|---------- |---------:|---------:|---------:|------:|--------:|
|   StdBool | 25.81 us | 0.211 us | 0.197 us |  1.00 |    0.00 |
|  FastBool | 11.24 us | 0.063 us | 0.059 us |  0.44 |    0.00 |
| OtherBool | 11.06 us | 0.224 us | 0.640 us |  0.45 |    0.02 |
````
OtherBool was significantly faster than StdBool, and virtually the same as FastBool

The SByte comparison code is similar to the Bool Comparison code above, except types as sbyte.
Results:
````
|     Method |      Mean |     Error |    StdDev | Ratio |
|----------- |----------:|----------:|----------:|------:|
|   StdSByte | 35.413 us | 0.6597 us | 0.6170 us |  1.00 |
|  FastSByte |  5.760 us | 0.0633 us | 0.0529 us |  0.16 |
| OtherSByte |  9.119 us | 0.0875 us | 0.0818 us |  0.26 |
````
Here, FastSByte is convincingly faster but OtherSByte is still a massive improvement and compatible with the rtm code.

Ditto for the Byte compare code.
````
|    Method |      Mean |     Error |    StdDev | Ratio |
|---------- |----------:|----------:|----------:|------:|
|   StdByte | 37.349 us | 0.7359 us | 1.1238 us |  1.00 |
|  FastByte |  5.897 us | 0.1075 us | 0.0953 us |  0.16 |
| OtherByte |  9.310 us | 0.1457 us | 0.1291 us |  0.25 |
````
Here, FastByte is convincingly faster but OtherByte is still a massive improvement and compatible with the rtm code.

````
|      Method |      Mean |     Error |    StdDev | Ratio |
|------------ |----------:|----------:|----------:|------:|
|   StdUInt16 | 35.496 us | 0.2151 us | 0.1906 us |  1.00 |
|  FastUInt16 |  6.014 us | 0.1196 us | 0.1753 us |  0.17 |
| OtherUInt16 |  9.384 us | 0.1842 us | 0.2460 us |  0.26 |
````
Here, FastUInt16 is faster but OtherUInt16 is still a massive improvement and compatible with the rtm code.

````
|     Method |      Mean |     Error |    StdDev | Ratio |
|----------- |----------:|----------:|----------:|------:|
|   StdInt16 | 35.982 us | 0.5151 us | 0.4566 us |  1.00 |
|  FastInt16 |  5.668 us | 0.0549 us | 0.0487 us |  0.16 |
| OtherInt16 |  9.109 us | 0.1464 us | 0.1369 us |  0.25 |
````
Ditto FastInt16 is faster but OtherInt16 is still a massive improvement and compatible with the rtm code.

My strong recommendation is that we go with the compatible implementations of these smaller data types, they are at least 3 times as fast as the current RTM release and are not compatible with the existing branched implementation.

thinkbeforecoding commented:
I understand.
I just thought that since byte.CompareTo / Int16.CompareTo etc were implemented like this, that it would not cause problems.

Reverting to using cgt - clt.

KevinRansom commented:
@thinkbeforecoding , the subtract only code is good,  the extra perf is very nice.   However,  just going branchless gives us a large performance bonus.  And not breaking existing code, makes it a no brainer to take.  Causing a possible behaviour change puts us in the position of having to decide whether to take it.  For sure we tend to be biased towards not breaking existing apps if we can.

This is a great change,  thank you for taking care of it mate.

thinkbeforecoding commented:
I also filed an issue for the jit to optimize both the useless double cmp and the zero ext which could be factored out.

thinkbeforecoding commented:
All systems go!

dsyme commented on tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare01.fsx.il.bsl:
```diff
@@ -69,39 +69,45 @@
       IL_0008:  ldc.i4.1
       IL_0009:  ldc.i4.1
       IL_000a:  cgt
-      IL_000c:  stloc.2
```
The increase in code size here has me a little perplaxed.  I'm assuming this is optimized code.  I'm not sure why the `cgt` hasn't been statically computed.

thinkbeforecoding replied:
The optimizer is matching on cgt when result is bool. I'm adding a new rule to return 1/0 when result is int32.
Same thing for clt. And we can add the same thing for any integer type.

dsyme commented on tests/FSharp.Compiler.ComponentTests/EmittedIL/GenericComparison/Compare02.fsx.il.bsl:
```diff
@@ -70,50 +70,58 @@
-
-      IL_0020:  ldc.i4.m1
+      IL_000c:  ldc.i4.0
+      IL_000d:  sub
```
The subtraction of zero again has be a little perplexed.  Any idea what's going on here?

dsyme commented on tests/fsharp/core/printing/output.stderr.txt:
```diff
@@ -0,0 +1,348 @@
+
```
None of these *.txt should be added in this PR, they should be added to `.gitignore`

dsyme reviewed (changes requested):
Please remove *.txt and *.dll from the PR

Have left a query about the case which gives larger codegen, we should look into that I think?

thinkbeforecoding commented:
And changes have been applied

thinkbeforecoding commented:
This time files have been removed from the commit directly (the dlls had been added prior to my commits. They are now deleted and I added them to gitignore)

thinkbeforecoding commented:
@vzarytovskii ... i don't have enough rights to rerun the build.. (the agent did not connect...)

vzarytovskii commented:
> @vzarytovskii ... i don't have enough rights to rerun the build.. (the agent did not connect...)

Huh, interesting. I didn't know you have to have any additional permissions. An easier way of kicking in the pipeline is to close and reopen the PR.

thinkbeforecoding commented:
When I used the command it said:
Commenter does not have sufficient privileges for PR 13187 in repo dotnet/fsharp

thinkbeforecoding commented:
There are still some cancelled builds... strange

thinkbeforecoding commented:
These canceled builds again.

KevinRansom reviewed (approved):
Nice, looks good.  Thanks

vzarytovskii commented:
> These canceled builds again. 

Yeah, some infra hiccups, I suppose.

thinkbeforecoding commented:
@dsyme , I think this just needs your review, but should be good.

KevinRansom commented:
@dsyme, this is good to go, do you want to review your change request.

dsyme commented:
@thinkbeforecoding Thank you so much for this improvement!

forki commented:
Wohooo! It landed. Awesome work!

EgorBo commented:
NOTE: as of .NET 7.0 (or was it 8.0), JIT is expected to use branchless instructions even when F#/C# emits branches (recognizes cmov like idioms)

thinkbeforecoding commented:
Yes I've noted this in some tests in fasmi. I think it's in dotnet 8.0

EgorBo commented:
Sure, I just wanted to note that it's better when some sub-optimal codegen is filed against JIT rather than silently (for JIT team) fixed 🙂
